### DISCUSSION TITLE
[General] support eth_estimateGas from a smart contract

### DISCUSSION THREAD

**Discussion by chendatony31** [zkSync-Community-Hub/zksync-developers#144] [General] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

Support eth_estimateGas from a smart contract

### What is the specific issue or error you're encountering?

here is a problem, when we are trying to call eth_estimateGas with from is a smart contract account, the rpc returns `invalid sender. can't start a transaction from a non-account`. is it prohibited by zksync by default?  As We (TokenPocket) supported the Safe's multi-sig wallet as a normal wallet to the dapps. And when the dapp tries to estimate and submit a transaction to the wallet, they get the error. Do we have a way to make the estimate work?

### Can you share the error messages or logs you're receiving, if any?

invalid sender. can't start a transaction from a non-account

### Have you made any recent changes to the contract before encountering this issue?

no

### Are there any external libraries or contracts that your contract interacts with?

no

### Can you provide the relevant portions of your contract code where the issue is occurring?

no

### Have you tried to isolate the problem, and if so, what were the results?

no

### What steps have you already taken to try to resolve the issue?

no

**Comment by dutterbutter** [comment]
Hey 👋 @chendatony31 

Can you provide a reproducible example or link a repo we can investigate?

**Comment by chendatony31** [comment]
Hi @dutterbutter 

Here is an example of rpc request `curl -H "Content-Type: application/json" https://zksync2-mainnet.zksync.io/ -X POST -d '{"id":8974198583840,"jsonrpc":"2.0","method":"eth_estimateGas","params":[{"from":"0xf2c3cbc6c63d03323b70f43bf13440bede395a27","to":"0x3355df6d4c9c3035724fd0e3914de96a5a83aaf4","value":"0x0","data":"0x095ea7b300000000000000000000000065de59507caa235c8ca49bfd008cada4d846caa7ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"}]}'` you can try it.

**Reply by dutterbutter** [reply]
The method you're attempting to use for `eth_estimateGas` isn't accurate. The curl request provided behaves as it is intended. To properly estimate gas for a function within your contract, you should use the following format:

```javascript
await contract.estimateGas.setGreeting("Your Argument Here");
```

**Reply by chendatony31** [reply]
The curl request is just a simple way to explain the request and params, no matter what method you are using finally it will become an HTTP request, no?  And it works for every EVM chains except zkSync.

**Comment by Zergity** [comment]
> The curl request provided behaves as it is intended.

No it's not. The same request works with any other EVM-compatible chains. This is clearly unexplained restriction by zkSync.

**Reply by dutterbutter** [reply]
Other EVM-compatible chains do not have native account abstraction which is the nuance here.

**Comment by katspaugh** [comment]
We're having the same issue. It works on all other chains except for zkSync.
https://github.com/safe-global/safe-wallet-web/issues/2631

**Comment by dutterbutter** [comment]
I see I overlooked the detail that the estimate call is being requested from a smart account. Let me review again and I will update my findings in this thread.

**Reply by ibwei** [reply]
any updates? I got the same issue.

**Reply by DefiDebauchery** [reply]
Adding another request for an update; this is hindering our ability to validate Timelock calldata (through our Safe, echoing kat's mention earlier)

**Reply by mshakeg** [reply]
I am also experiencing the same issue. In my case, I'm attempting to estimate gas for transactions initiated by a Gnosis Safe contract on the zkSync Era network. When using `eth_estimateGas` with the `from` field set to a smart contract address (specifically, a Gnosis Safe), I encounter the error "invalid sender. can't start a transaction from a non-account." On networks like the Ethereum mainnet estimating gas in this manner works without issues.

**Reply by jt892** [reply]
I got same issue as people here. After read your comment, I tried to estimate gas (I filled 25000), LOL, the transaction done. Thank you.

**Reply by uF4No** [reply]
hey @jt892 can you share the tx id in which you hardcoded the gas and that was executed successfully?

**Reply by jt892** [reply]
Of course. 0xd94aa1ff620ec57fd333e97c7976c2dfcc9d144efeb26920543b5ad2d2a20960

**Reply by uF4No** [reply, marked as answer]
@jt892  I think that's a different scenario as the `from` in that transaction is an EOA, and the problem reported here is when the `from` in the tx is another Safe. 
In those cases, looks like the problem is that Safe accounts are not deployed as native smart contract accounts so zkSync returns the error `invalid sender. can't start a transaction from a non-account`. 

To summarise, zkSync has native account abstraction and, under the hood, all accounts are a smart contract account. Even EOA use the [DefaultAccount](https://github.com/matter-labs/era-system-contracts/blob/main/contracts/DefaultAccount.sol) smart contract. All smart contract accounts on zkSync must be deployed using the `createAccount` or `create2Account` methods of the [`ContractDeployer` system contract](https://github.com/matter-labs/era-system-contracts/blob/main/contracts/ContractDeployer.sol).

When processing a transaction, the protocol checks the code of the `from` account and, in this case, as Safe accounts are not deployed as native accounts on zkSync (via `createAccount` or `create2Account`), it fails with the error above. 

For now, I'd avoid creating Safe accounts owned by another Safe as the only way to support this scenario is via protocol changes or by changing the way Safe accounts are deployed on zkSync.

**Reply by jjm340** [reply]
Hi, if you do end up setting up a safewallet with another safewallet as owner, and you have some funds in the account, what are your options?

**Comment by candacekrank** [comment]
It seems like you're encountering an issue when trying to use eth_estimateGas with a smart contract account as the sender. This error message typically indicates that the sender address provided is not recognized as a valid Ethereum account.

In the context of zkSync, transactions are typically initiated from externally owned accounts (EOAs) rather than smart contracts. It's possible that zkSync does not support transactions initiated from smart contract accounts by default, which could explain the error message you're receiving.

To resolve this issue, you may need to modify your application logic to ensure that transactions are initiated from EOAs rather than smart contracts. Alternatively, you can explore if zkSync provides any specific mechanisms or APIs for interacting with smart contract accounts in this manner.

If you're unsure, you can reach out to the zkSync team or consult their documentation for further guidance on this matter.

**Comment by sarportzapbisusan** [comment]
It seems that the issue arises when attempting to call eth_estimateGas with a smart contract account as the sender. This results in an "invalid sender" error, indicating that transactions cannot be initiated from a non-account. This behavior might be inherent to zkSync's design, preventing transactions from being started by non-account entities such as smart contracts. Further investigation and potential modifications to the design or implementation may be needed to address this issue.

**Comment by lykinsjennifer** [comment]
Currently, zksync does not support calling eth_estimateGas from a smart contract account. Transactions must be initiated from an externally owned account (EOA). This limitation may affect dapps interacting with smart contracts, such as multi-sig wallets like Safe, when estimating gas for transactions. One potential workaround could be to estimate gas locally within the dapp before submitting transactions to the wallet.

**Comment by colleennealv** [comment]
Transactions initiated from smart contract accounts are not supported by zkSync. Transactions must originate from externally owned accounts (EOAs) or accounts controlled by zkSync wallets. To estimate gas for transactions involving smart contracts, you may need to refactor your implementation to use an EOA as the transaction sender.

**Comment by knowenlophoroger** [comment]
The user is encountering an issue when trying to call eth_estimateGas from a smart contract account on zksync mainnet, as it returns an "invalid sender" error. The user wants to know if this is prohibited by zksync by default, and if there is a way to make the estimate work. No relevant portions of contract code are provided, and the user has not attempted to isolate or resolve the issue yet. No repository link or additional details are provided.
